Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -0,0 +1,124 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Computability.Language
+
+/-!
+# Context-Free Grammar
+
+This file contains the definition of a Context-Free Grammar, which is a grammar that has a single
+nonterminal symbol on the left-hand side of each rule.
+-/
+
+/-- Rule that rewrites a single nonterminal to any list of symbols. -/
+structure ContextFreeRule (T : Type*) (N : Type*) where
+  /-- Input nonterminal a.k.a. left-hand side -/
+  input : N
+  /-- Output string a.k.a. right-hand side -/
+  output : List (Symbol T N)
+
+/-- Context-free grammar that generates words over the alphabet `T` (a type of terminals). -/
+structure ContextFreeGrammar (T : Type*) where
+  /-- Type of nonterminals -/
+  NT : Type*
+  /-- Initial nonterminal -/
+  initial : NT
+  /-- Rewrite rules -/
+  rules : List (ContextFreeRule T NT)
+
+variable {T : Type*}
+
+/-- One application of single context-free rule. -/
+inductive ContextFreeRule.RewritesTo {N : Type*} (r : ContextFreeRule T N) :
```

Comment:
this name isn't great but I can't think of a better one so; it'd be nice if you could have something like `r.Rewrites p to q` but alas

Reply:
I renamed `RewritesTo` to `Rewrites` to make it less ugly.

Reply:
PS: Thank god for the gray background behind code blocks on GitHub!